spark/src/main/java/org/apache/iceberg/actions/PlanScanAction.java

```diff
@@ -0,0 +1,263 @@
+          scan.table(), snapshot.snapshotId(), snapshot.timestampMillis(), scan.filter());
+      Listeners.notifyAll(
+          new ScanEvent(scan.table().toString(), snapshot.snapshotId(), scan.filter(), scan.schema()));
+      return planFilesFor();
+    } else {
+      LOG.info("Scanning empty table {}", scan.table());
+      return CloseableIterable.empty();
+    }
+  }
+
+  private CloseableIterable<FileScanTask> planFilesFor() {
+    // This is different for different BaseTableScan implementations
+    if (scan instanceof DataTableScan) {
+      return planDataTableScanFiles();
+    } else {
+      throw new UnsupportedOperationException(String.format("Unable to do distributed planning for %s",
+          scan.getClass()));
+    }
+  }
+
+  private DeleteFileIndex buildDeleteFileIndex() {
+    // Build delete index locally
+    List<ManifestFile> deleteManifests = snapshot.deleteManifests();
+    DeleteFileIndex.Builder deleteFileIndexBuilder = DeleteFileIndex.builderFor(table.io(), deleteManifests);
+    deleteFileIndexBuilder.caseSensitive(scan.isCaseSensitive());
+    deleteFileIndexBuilder.specsById(table.specs());
+    deleteFileIndexBuilder.filterData(scan.filter());
+    return deleteFileIndexBuilder.build();
+  }
+
+  private CloseableIterable<FileScanTask> planDataTableScanFiles() {
+    // TODO Currently this approach reads all manifests, no manifest filtering - Maybe through pushdowns or table read
+    // options?
+
+
+    // Create a dataframe of all DataFile entries
+    String dataFilesMetadataTable = metadataTableName(MetadataTableType.ENTRIES);
+    Dataset<Row> manifestEntries =
+        spark.read()
+            .format("iceberg")
+            .load(dataFilesMetadataTable);
+
+    // Todo pushdown filters to ManifestEntriesTable
+    // Read entries which are not deleted and are datafiles and not delete files
+    Dataset<Row> dataFileEntries = manifestEntries
+        .filter(manifestEntries.col("data_file").getField(DataFile.CONTENT.name()).equalTo(0)) // Only DataFiles
+        .filter(manifestEntries.col("status").notEqual(2)); // No Deleted Files
+
+    // Handle Incremental Scans
+    if (scan instanceof IncrementalDataTableScan) {
+      IncrementalDataTableScan incScan = (IncrementalDataTableScan) scan;
+      List<Snapshot> snapshots = IncrementalDataTableScan.snapshotsWithin(table, incScan.fromSnapshotId(),
+          incScan.toSnapshotId());
+      List<Long> validSnapshotIds = snapshots.stream().map(Snapshot::snapshotId).collect(Collectors.toList());
+      dataFileEntries = dataFileEntries
+          .filter(dataFileEntries.col("snapshot_id").isin(validSnapshotIds.toArray()));
+    }
+
+    // Build up evaluators and filters for Metrics and Partition values
+    final Expression scanFilter = scan.filter();
+    final boolean ignoreResiduals = scan.shouldIgnoreResiduals();
+    final boolean isCaseSensitive = scan.isCaseSensitive();
```

Comment:
I know there's debate about it theoretically allowing the compiler to make better choices, but we usually don't declare local variables final.

Reply:
ah yeah, this is me refactoring things which use to be class fields to method variables and not changing the attributes